Add ARC MPU v3 & v6 support

YuguoWH opened the pull request:
Adding ARC MPU v3 and v6 support. 
The v3 uses same logic with v2 and the only difference is to support 32 bytes minimal region size. 
The v6 support 32 bytes, and it can have up to 32 regions. The regions are divided into 2 banks and switched by set/unset RB bit in MPU_EN aux register
This PR also introduces a new nsim hs board simulator which can validate/test with mpu v6.

YuguoWH committed:
arch: arc: add support of mpu v3

Add support of ARC mpu version 3 which can have region size down to 32
bytes

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

YuguoWH commented:
related issue https://github.com/zephyrproject-rtos/zephyr/issues/37309

abrodkin commented on include/arch/arc/arch.h:
```diff
@@ -67,6 +67,8 @@ extern "C" {
 #elif CONFIG_ARC_MPU_VER == 4
 #define Z_ARC_MPU_ALIGN	32
+#elif CONFIG_ARC_MPU_VER == 6
+#define Z_ARC_MPU_ALIGN 32
```
Maybe
```
#if (CONFIG_ARC_MPU_VER == 4) || (CONFIG_ARC_MPU_VER == 6)
```
?

abrodkin commented on include/arch/arc/v2/linker.ld:
```diff
@@ -32,6 +32,8 @@
 	#elif CONFIG_ARC_MPU_VER == 4
 		#define MPU_MIN_SIZE 32
+	#elif CONFIG_ARC_MPU_VER == 6
+		#define MPU_MIN_SIZE 32
```
Ditto

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright (c) 2019 Synopsys.
```
2021 ;)

abrodkin commented on arch/arc/core/mpu/arc_mpu_v2_internal.h:
```diff
@@ -17,7 +17,12 @@
+#if CONFIG_ARC_MPU_VER == 2
 #define ARC_FEATURE_MPU_ALIGNMENT_BITS 11
+/* For MPU version 3, the minimum protection region size is 32 bytes */
+#else
```
Maybe do explicit check for versions 3 & 6 to get rid of ambiguity?

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,455 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V2_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V2_INTERNAL_H_
```
`ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_` (see **v6**)

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,455 @@
+	 * configured regions. The type is one-indexed rather than
+	 * zero-indexed.
+	 *
+	 * For ARC MPU v2, the smaller index has higher priority, so the
```
Why mention MPUv2 in code for MPUv6?

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,455 @@
+	int r_index;
+
+	/*
+	 * For ARC MPU v2, smaller region number takes priority.
```
Ditto

abrodkin reviewed (changes requested):
Maybe move common functionality into version-agnostic file and then do specific things in v2 & v6 files to reduce duplication of code?

Also see some minor comments.

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,455 @@
+	r_addr_end = r_addr_start  + (1 << (r_size_lshift + 1));
+
+	if (start >= r_addr_start && (start + size) <= r_addr_end) {
+		return 1;
```
return true;

evgeniy-paltsev reviewed (changes requested):
Hi Yuguo,

please find some comments (mostly minor).

YuguoWH committed:
arch: arc: add support of mpu v6

Add support of ARC mpu v6
* minimal region size down to 32 bytes
* maximal region number up to 32
* not support uncacheable region and volatile uncached region
* clean up mpu code for better readablity

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+	uint32_t index = (r_index % ARC_FEATURE_MPU_BANK_SIZE) * 2U;
+
+	_bank_select(bank);
+	return ((z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
```
Maybe explicitly return some `bool`?

evgeniy-paltsev replied:
We are returning bool here (two values comparation result) but it is hidden as we split it for two lines...

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+
+#define AUX_MPU_RDP_REGION_SIZE(bits)   (((bits - 1) & 0x3) | (((bits - 1) & 0x1C) << 7))
+
+#define AUX_MPU_RDP_ATTR_MASK (0x1F8)
```
Frankly I don't like these cryptic numbers. I guess these are some groups of bits shifted left?
Look at `BIT_MASK()` in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/sys/util_macro.h#L68. Then we'll get something like `BIT_MASK(xxx) << yyy` which is much more readable.

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
+#define AUX_MPU_EN_ENABLE   (0x40000000)
```
Maybe `BIT(x)` and `~BIT(x)` correspondigly?

abrodkin commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
```
`BIT(0)`?

abrodkin reviewed (approved):
With a couple of style nitpicks LGTM.

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_common_internal.h:
```diff
@@ -0,0 +1,300 @@
+	 * matched region that grants permission or denies access.
+	 *
+	 */
+	for (r_index = 0; r_index < get_num_regions(); r_index++) {
```
As we use r_index only in loop probably it's better to define it in loop itself:
```
for (int r_index = 0; r_index < get_num_regions(); r_index++) {
```

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_common_internal.h:
```diff
@@ -0,0 +1,300 @@
+	/* Disable MPU */
+	arc_core_mpu_disable();
+
+	int r_index;
+	/*
+	 * the MPU regions are filled in the reverse order.
+	 * According to ARCv2 ISA, the MPU region with smaller
+	 * index has higher priority. The static background MPU
+	 * regions in mpu_config will be in the bottom. Then
+	 * the special type regions will be above.
+	 *
+	 */
+	r_index = num_regions - mpu_config.num_regions;
```
I don't think we need to split definition and declaration here. Let's combine it:
```
/* comment */
int r_index = num_regions - mpu_config.num_regions;
```

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+	 */
+	switch (type) {
+	case THREAD_STACK_USER_REGION:
+		return get_num_regions() - mpu_config.num_regions
+		       - THREAD_STACK_REGION;
```
We don't exceed the Zephyr line limit here (100 symbols per line), please keep it in one line.

Please check rest of the code - we have a lot of places where we can put the whole expression in one line. It improves code readability a lot.

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+	uint32_t index = (r_index % ARC_FEATURE_MPU_BANK_SIZE) * 2U;
+
+	_bank_select(bank);
+	r_addr_start = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
+		       & (~AUX_MPU_RDB_VALID_MASK);
```
Ditto.

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+	_bank_select(bank);
+	r_addr_start = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDB0 + index)
+		       & (~AUX_MPU_RDB_VALID_MASK);
+	r_size_lshift = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDP0 + index)
+			& AUX_MPU_RDP_SIZE_MASK;
```
Ditto.

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_v6_internal.h:
```diff
@@ -0,0 +1,162 @@
+		       & (~AUX_MPU_RDB_VALID_MASK);
+	r_size_lshift = z_arc_v2_aux_reg_read(_ARC_V2_MPU_RDP0 + index)
+			& AUX_MPU_RDP_SIZE_MASK;
+	r_size_lshift = (r_size_lshift & 0x3) | ((r_size_lshift >> 7) & 0x1C);
```
Should we move those magic numbers to some defines / constants?

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_common_internal.h:
```diff
@@ -0,0 +1,300 @@
+	r_index = num_regions - mpu_config.num_regions;
+
+	/* clear all the regions first */
+	for (i = 0U; i < r_index; i++) {
```
As we use `i` only in loop probably it's better to define it in loop itself:
```
for (uint32_t i = 0U; i < r_index; i++) {
```

evgeniy-paltsev commented on arch/arc/core/mpu/arc_mpu_common_internal.h:
```diff
@@ -0,0 +1,300 @@
+	}
+
+	/* configure the static regions */
+	for (i = 0U; i < mpu_config.num_regions; i++) {
```
Ditto.

abrodkin reviewed (changes requested):
In commit titles it's better to say "add support _of_" instead of "to".
Also some comments below.

YuguoWH committed:
boards: arc: add a nsim_hs_mpuv6 board simulator

We add support of mpu v6 therefore it is needed to have a board to
validate that feature. This commit add a new HS nsim simulator
which supports mpu v6.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

abrodkin reviewed (approved):
With 2 nitpicks LGTM

ruuddw commented on soc/arc/snps_nsim/Kconfig:
```diff
@@ -29,6 +29,11 @@ config SOC_NSIM_HS_SMP
+config SOC_NSIM_HS_MPUV6
+	bool "Synopsys ARC HS with MPU v6 in nSIM"
+	select CPU_HAS_MPU
+	select CPU_HAS_FPU
```
Why is FPU enabled?

YuguoWH replied:
I keep it same with other NSIM_HS boards, so FPU is enabled
